File: 438-find-all-anagrams-in-a-string/find-all-anagrams-in-a-string.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        unordered_map<char, int> mp;
        for(char ch : p)
            mp[ch]++;
        
        int n1 = s.size(), n2 = p.size();
        vector<int> ans;
        for(int i = 0; i <= n1-n2; i++){
            int j = i;
            unordered_map<char, int> anaFreq;
            while(j-i < n2){
                anaFreq[s[j]]++;
                j++;
            }
            if(mp == anaFreq)
                ans.push_back(i);
        }
        return ans;
    }
};

File: 438-find-all-anagrams-in-a-string/find-all-anagrams-in-a-string.java
class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        List<Integer> ans = new ArrayList<>();

        int[] freq = new int[26];

        for(char ch : p.toCharArray())
            freq[ch - 'a']++;

        int l = 0;

        for(int h = 0; h < s.length(); h++){
            freq[s.charAt(h) - 'a']--;

            while(freq[s.charAt(h) - 'a'] < 0){
                freq[s.charAt(l) - 'a']++;
                l++;
            }
            if(h - l + 1 == p.length()){
                ans.add(l);
            }
        }

        return ans;
    }
}